File: src/vrtgen/types/prologue.py
# Copyright (C) 2019 Geon Technologies, LLC
#
# This file is part of vrtgen.
#
# vrtgen is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option)
# any later version.
#
# vrtgen is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
# more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

"""
Types used for VITA 49 packet prologues.
"""

from .struct import Struct, Field, Reserved
from . import basic, enums

class ClassIdentifier(Struct):
    """
    Class Identifier (5.1.3).
    """
    pad_bits = Field('Pad Bit Count', basic.IntegerType.create(5, signed=False))
    reserved_1 = Reserved(3)
    oui = Field('Organizationally Unique Identifier', basic.OUI)
    information_code = Field('Information Class Code', basic.Identifier16)
    packet_code = Field('Packet Class Code', basic.Identifier16)

class Header(Struct):
    """
    VRT Packet Header (5.1.1).
    """
    packet_type = Field('Packet Type', enums.PacketType)
    class_id_enable = Field('Class Identifier', basic.Boolean)
    packet_specific_26 = Reserved(1)
    packet_specific_25 = Reserved(1)
    packet_specific_24 = Reserved(1)
    tsi = Field('TSI', enums.TSI)
    tsf = Field('TSF', enums.TSF)
    packet_count = Field('Packet Count', basic.IntegerType.create(4, signed=False))
    packet_size = Field('Packet Size', basic.UInteger16)

class DataHeader(Header):
    """
    VRT Packet Header with Data Packet-Specific Indicator Bits (5.1.1.1)
    """
    trailer_included = Header.packet_specific_26.rebind(
        Field('Trailer Included', basic.Boolean)
    )
    not_v49d0 = Header.packet_specific_25.rebind(
        Field('Not a V49d0 Packet', basic.Boolean)
    )
    spectrum = Header.packet_specific_24.rebind(
        Field('Signal Spectrum or Signal Time Data Packet', basic.Boolean)
    )

class ContextHeader(Header):
    """
    VRT Packet Header with Context Packet-Specific Indicator Bits (5.1.1.1)
    """
    not_v49d0 = Header.packet_specific_25.rebind(
        Field('Not a V49d0 Packet', basic.Boolean)
    )
    timestamp_mode = Header.packet_specific_24.rebind(
        Field('Timestamp Mode', enums.TSM)
    )

class CommandHeader(Header):
    """
    VRT Packet Header with Command Packet-Specific Indicator Bits (5.1.1.1)
    """
    acknowledge_packet = Header.packet_specific_26.rebind(
        Field('Acknowledge Packet', basic.Boolean)
    )
    cancellation_packet = Header.packet_specific_24.rebind(
        Field('Cancellation Packet', basic.Boolean)
    )

class Prologue(Struct):
    """
    VRT Packet Prologue (5).
    """
    stream_id = Field('Stream Identifier', basic.StreamIdentifier)
    class_id = Field('Class Identifier', ClassIdentifier)
    integer_timestamp = Field('Integer Timestamp', basic.UInteger32)
    fractional_timestamp = Field('Fractional Timestamp', basic.UInteger64)

File: cpp/tests/libvrtgen/prologue.cpp
/*
 * Copyright (C) 2019 Geon Technologies, LLC
 *
 * This file is part of vrtgen.
 *
 * vrtgen is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * vrtgen is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */


#include "catch/catch.hpp"

#include "vrtgen/enums.hpp"
#include "vrtgen/packing/cif0.hpp"
#include "vrtgen/packing/cif1.hpp"
#include "vrtgen/packing/header.hpp"
#include "vrtgen/packing/trailer.hpp"

#include "bytes.hpp"

TEST_CASE("Header getter methods")
{
    using vrtgen::packing::Header;

    bytes data = {
        0x10, 0x5c, 0x12, 0x34,
    };
    Header& header = *reinterpret_cast<Header*>(data.data());
    SECTION("Packet Type") {
        REQUIRE(header.getPacketType() == vrtgen::PacketType::SIGNAL_DATA_STREAM_ID);
    }
    SECTION("Class Identifier Enable") {
        REQUIRE(header.isClassIdentifierEnabled() == false);
    }
    SECTION("TSI") {
        REQUIRE(header.getTSI() == vrtgen::TSI::UTC);
    }
    SECTION("TSF") {
        REQUIRE(header.getTSF() == vrtgen::TSF::SAMPLE_COUNT);
    }
    SECTION("Packet Count") {
        REQUIRE(header.getPacketCount() == 12);
    }
    SECTION("Packet Size") {
        REQUIRE(header.getPacketSize() == 0x1234);
    }
}

TEST_CASE("Header setter methods")
{
    using vrtgen::packing::Header;

    bytes data = {0, 0, 0, 0}; // zero-initialize
    Header& header = *reinterpret_cast<Header*>(data.data());

    SECTION("Packet Type") {
        header.setPacketType(vrtgen::PacketType::CONTEXT);
        REQUIRE(data == bytes({0x40, 0x00, 0x00, 0x00}));
    }
    SECTION("Class Identifier Enable") {
        header.setClassIdentifierEnabled(true);
        REQUIRE(data == bytes({0x08, 0x00, 0x00, 0x00}));
    }
    SECTION("TSI") {
        header.setTSI(vrtgen::TSI::GPS);
        REQUIRE(data == bytes({0x00, 0x80, 0x00, 0x00}));
    }
    SECTION("TSF") {
        header.setTSF(vrtgen::TSF::REAL_TIME);
        REQUIRE(data == bytes({0x00, 0x20, 0x00, 0x00}));
    }
    SECTION("Packet Count") {
        header.setPacketCount(12);
        REQUIRE(data == bytes({0x00, 0x0C, 0x00, 0x00}));
    }
    SECTION("Packet Size") {
        header.setPacketSize(0xCDEF);
        REQUIRE(data == bytes({0x00, 0x00, 0xCD, 0xEF}));
    }
}

TEST_CASE("Data Header getter methods")
{
    using vrtgen::packing::DataHeader;

    bytes data = {0, 0, 0, 0};
    DataHeader& header = *reinterpret_cast<DataHeader*>(data.data());

    SECTION("Trailer Included") {
        data[0] = 0x04;
        REQUIRE(header.getTrailerIncluded());
    }
    SECTION("Not a V49.0 Packet") {
        data[0] = 0x02;
        REQUIRE(header.getNotaV49_0Packet());
    }
    SECTION("Signal Spectrum or Signal Time Data Packet") {
        data[0] = 0x01;
        REQUIRE(header.getSignalSpectrumorSignalTimeDataPacket());
    }
}

TEST_CASE("Data Header setter methods")
{
    using vrtgen::packing::DataHeader;

    bytes data = {0, 0, 0, 0};
    DataHeader& header = *reinterpret_cast<DataHeader*>(data.data());

    SECTION("Trailer Included") {
        header.setTrailerIncluded(true);
        REQUIRE(data == bytes({0x04, 0x00, 0x00, 0x00}));
    }
    SECTION("Not a V49.0 Packet") {
        header.setNotaV49_0Packet(true);
        REQUIRE(data == bytes({0x02, 0x00, 0x00, 0x00}));
    }
    SECTION("Signal Spectrum or Signal Time Data Packet") {
        header.setSignalSpectrumorSignalTimeDataPacket(true);
        REQUIRE(data == bytes({0x01, 0x00, 0x00, 0x00}));
    }
}

TEST_CASE("Context Header getter methods")
{
    using vrtgen::packing::ContextHeader;

    bytes data = {0, 0, 0, 0};
    ContextHeader& header = *reinterpret_cast<ContextHeader*>(data.data());

    SECTION("Not a V49.0 Packet") {
        CHECK_FALSE(header.getNotaV49_0Packet());
        data[0] = 0x02;
        REQUIRE(header.getNotaV49_0Packet());
    }
    SECTION("Timestamp Mode") {
        CHECK(header.getTimestampMode() == vrtgen::TSM::FINE);
        data[0] = 0x01;
        CHECK(header.getTimestampMode() == vrtgen::TSM::COARSE);
    }
}

TEST_CASE("Context Header setter methods")
{
    using vrtgen::packing::ContextHeader;

    bytes data = {0, 0, 0, 0};
    ContextHeader& header = *reinterpret_cast<ContextHeader*>(data.data());

    SECTION("Not a V49.0 Packet") {
        header.setNotaV49_0Packet(true);
        REQUIRE(data == bytes({0x02, 0x00, 0x00, 0x00}));
    }
    SECTION("Timestamp Mode") {
        header.setTimestampMode(vrtgen::TSM::COARSE);
        CHECK(data == bytes({0x01, 0x00, 0x00, 0x00}));
        header.setTimestampMode(vrtgen::TSM::FINE);
        CHECK(data == bytes({0x00, 0x00, 0x00, 0x00}));
    }
}

TEST_CASE("Command Header getter methods")
{
    using vrtgen::packing::CommandHeader;

    bytes data = {0, 0, 0, 0};
    CommandHeader& header = *reinterpret_cast<CommandHeader*>(data.data());

    SECTION("Acknowledge") {
        CHECK_FALSE(header.getAcknowledgePacket());
        data[0] = 0x04;
        CHECK(header.getAcknowledgePacket());
    }
    SECTION("Cancellation") {
        CHECK_FALSE(header.getCancellationPacket());
        data[0] = 0x01;
        CHECK(header.getCancellationPacket());
    }
}

TEST_CASE("Command Header setter methods")
{
    using vrtgen::packing::CommandHeader;

    bytes data = {0, 0, 0, 0};
    CommandHeader& header = *reinterpret_cast<CommandHeader*>(data.data());

    SECTION("Acknowledge") {
        header.setAcknowledgePacket(true);
        REQUIRE(data == bytes({0x04, 0x00, 0x00, 0x00}));
    }
    SECTION("Cancellation") {
        header.setCancellationPacket(true);
        REQUIRE(data == bytes({0x01, 0x00, 0x00, 0x00}));
    }
}

TEST_CASE("ClassIdentifier getter methods")
{
    using vrtgen::packing::ClassIdentifier;

    bytes data = {0x28, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF};
    ClassIdentifier& class_id = *reinterpret_cast<ClassIdentifier*>(data.data());
    CHECK(sizeof(class_id) == data.size());

    SECTION("Pad Bit Count") {
        REQUIRE(class_id.getPadBitCount() == 5);
    }
    SECTION("Organizationally Unique Identifier") {
        REQUIRE(class_id.getOrganizationallyUniqueIdentifier() == 0x234567);
    }
    SECTION("Information Class Code") {
        REQUIRE(class_id.getInformationClassCode() == 0x89AB);
    }
    SECTION("Packet Class Code") {
        REQUIRE(class_id.getPacketClassCode() == 0xCDEF);
    }
}

TEST_CASE("ClassIdentifier setter methods")
{
    using vrtgen::packing::ClassIdentifier;

    bytes data = {0, 0, 0, 0, 0, 0, 0, 0};
    ClassIdentifier& class_id = *reinterpret_cast<ClassIdentifier*>(data.data());

    SECTION("Pad Bit Count") {
        class_id.setPadBitCount(0x13);
        REQUIRE(data == bytes({0x98, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}));
    }
    SECTION("Organizationally Unique Identifier") {
        class_id.setOrganizationallyUniqueIdentifier(0x5307C0);
        REQUIRE(data == bytes({0x00, 0x53, 0x07, 0xC0, 0x00, 0x00, 0x00, 0x00}));
    }
    SECTION("Information Class Code") {
        class_id.setInformationClassCode(0x916F);
        REQUIRE(data == bytes({0x00, 0x00, 0x00, 0x00, 0x91, 0x6F, 0x00, 0x00}));
    }
    SECTION("Packet Class Code") {
        class_id.setPacketClassCode(0x1234);
        REQUIRE(data == bytes({0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x12, 0x34}));
    }
}
